### PR TITLE
Allow Built in Resource Packs to have custom names

### PR DESCRIPTION
**Pull request by Jab125**

### CODE REVIEW

**Comment by Jab125** [comment]
I don't know how to bump the versions, so can somebody please bump the resource helper's version.

**Comment by modmuss50** [comment]
> I don't know how to bump the versions, so can somebody please bump the resource helper's version.

No need to bump versions in PRs, it will be done before release.

**Comment by Jab125** [comment]
ah

**Comment by Jab125** [comment]
its probably trailing whitespace

**Comment by Jab125** [comment]
i think its fixed

**Comment by Jab125** [comment]
@modmuss50 All good now, i managed to fix the trailing whitespace

**Comment by modmuss50** [comment]
Its still failing, if you run `gradlew build` locally to ensure that it passes the checkstyle rules. You can also get an IDE plugin if that suits you better.

**Comment by Jab125** [comment]
successful build :)

**Comment by Jab125** [comment]
I'm just gonna update the testmod

**Comment by Technici4n** [comment]
Hi, could you post before/after pictures?

**Comment by LambdAurora** [comment]
There's reasons I haven't implemented this originally:
 - this could make some resource packs very hard to locate from which mod it comes from
 - no localization, the issue also exists for the descriptions, and I think it's terrible
 - the description field should be used to describe the pack, not the name
 - this makes harder to determine which resource packs come from the game, which come from the resource pack folder, and which come from mods.

**Comment by Jab125** [comment]
> Hi, could you post before/after pictures?

don't really have pictures, but before it was one test pack and now there are two, one is named, one isn't

assets (resource pack) is in first one
data (data pack) is in second one


> There's reasons I haven't implemented this originally:
> 
> * this could make some resource packs very hard to locate from which mod it comes from
> * no localization, the issue also exists for the descriptions, and I think it's terrible
> * the description field should be used to describe the pack, not the name
> * this makes harder to determine which resource packs come from the game, which come from the resource pack folder, and which come from mods.

Any mod can pass in the wrong mod container or Identifier

it displays (built-in) in the description like this
[Image](https://cdn.discordapp.com/attachments/895859330094624790/909310662084284446/Screen_Shot_2021-11-14_at_4.15.52_pm.png)

File: fabric-resource-loader-v0/src/testmod/java/net/fabricmc/fabric/test/resource/loader/BuiltinResourcePackTestMod.java [hunk 1]
```diff
@@ -36,6 +36,10 @@ public void onInitialize() {
 		// Should always be present as it's **this** mod.
 		FabricLoader.getInstance().getModContainer(MODID)
 				.map(container -> ResourceManagerHelper.registerBuiltinResourcePack(new Identifier(MODID, "test"),
+						container, "Fabric Resource Loader Test Pack", ResourcePackActivationType.NORMAL))
+				.filter(success -> !success).ifPresent(success -> LOGGER.warn("Could not register built-in resource pack."));
+		FabricLoader.getInstance().getModContainer(MODID)
+				.map(container -> ResourceManagerHelper.registerBuiltinResourcePack(new Identifier(MODID, "test2"),
```
Comment (LambdAurora) [inline fabric-resource-loader-v0/src/testmod/java/net/fabricmc/fabric/test/resource/loader/BuiltinResourcePackTestMod.java:42]: This doesn't need new lambdas, you can put it in the original map with an and.

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/api/resource/ResourceManagerHelper.java [hunk 2]
```diff
@@ -75,6 +75,27 @@ static boolean registerBuiltinResourcePack(Identifier id, ModContainer container
...
+	 *
+	 * <p>Why and when to use it? A built-in resource pack should be used to provide extra assets/data that should be optional with your mod but still directly provided by it.
+	 * For example it could provide textures of your mod in another resolution, or could allow to provide different styles of your assets.
+	 *
+	 * <p>The path in which the resource pack is located is in the mod JAR file under the {@code "resourcepacks/<id path>"} directory. {@code id path} being the path specified
+	 * in the identifier of this built-in resource pack.
+	 *
+	 * @param id             the identifier of the resource pack
+	 * @param container      the mod container
+	 * @param displayName    the display name of the resource pack
```
Comment (LambdAurora) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/api/resource/ResourceManagerHelper.java]:
```suggestion
	 * @param displayName    the display name of the resource pack, should include mod name for clarity
```

**Review by LambdAurora — COMMENTED** [review]
If this goes through, it would be a good idea [to modify this](https://github.com/FabricMC/fabric/blob/f09604ce0ad1f8bf3e1bf055583af9c80db40723/fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java#L108) to use its own pack source (mod built-in), but it comes with the issue of needing to be translated.

Ultimately, this can be a good change, my only worry is mods trying to play sneaky with it.

**Comment by Jab125** [comment]
Gonna attempt to use I18n

**Comment by Jab125** [comment]
> If this goes through, it would be a good idea [to modify this](https://github.com/FabricMC/fabric/blob/f09604ce0ad1f8bf3e1bf055583af9c80db40723/fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java#L108) to use its own pack source (mod built-in), but it comes with the issue of needing to be translated.
> 
> Ultimately, this can be a good change, my only worry is mods trying to play sneaky with it.

It uses a new pack source now

**Comment by LambdAurora** [comment]
> It uses a new pack source now

Amazing, couldn't ask more!
To me this can go through like that, my wishes of localization may be out-of-scope for this PR.

Also, since a new translation key is introduced, for `fr_fr.json`, and `fr_ca.json`, this string should be added (I am French, and multiple people can confirm it):
```json
"pack.source.builtinMod": "intégré: %s"
```

**Comment by Jab125** [comment]
> > It uses a new pack source now
> 
> Amazing, couldn't ask more! To me this can go through like that, my wishes of localization may be out-of-scope for this PR.
> 
> Also, since a new translation key is introduced, for `fr_fr.json`, and `fr_ca.json`, this string should be added (I am French, and multiple people can confirm it):
> 
> ```json
> "pack.source.builtinMod": "intégré: %s"
> ```

Lang has been added

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java [hunk 3]
```diff
@@ -173,4 +193,9 @@ public void registerReloadListener(IdentifiableResourceReloadListener listener)
 			throw new RuntimeException("Listener with previously unknown ID " + listener.getFabricId() + " already in listener set!");
 		}
 	}
+
+	private static ResourcePackSource getResourceSource(String modid) {
+		String trimmedId = StringUtils.abbreviate(modid, 8);
```
Comment (Juuxel) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java]:
```suggestion
	private static ResourcePackSource getResourceSource(String modId) {
		String trimmedId = StringUtils.abbreviate(modId, 8);
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

**Review by Juuxel — APPROVED** [review]

**Review by deirn — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]

**Review by Technici4n — APPROVED** [review]

**Review by Player3324 — APPROVED** [review]

**Comment by Technici4n** [comment]
Note: closes #1138.

**Review by LambdAurora — CHANGES_REQUESTED** [review]
This pull request as-is will break built-in resource packs that are not enabled by default on servers as they will be made loading by default.
Reason being the mixin injection relies on reference equality of the source of the resource pack.

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java [hunk 4]
```diff
@@ -173,4 +193,9 @@ public void registerReloadListener(IdentifiableResourceReloadListener listener)
 			throw new RuntimeException("Listener with previously unknown ID " + listener.getFabricId() + " already in listener set!");
 		}
 	}
+
+	private static ResourcePackSource getResourceSource(String modId) {
+		String trimmedId = StringUtils.abbreviate(modId, 8);
```
Comment (Technici4n) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java]: Not sure trimming the modid is a good idea.

**Comment by deirn** [comment]
> This pull request as-is will break built-in resource packs that are not enabled by default on servers as they will be made loading by default. Reason being the mixin injection relies on reference equality of the source of the resource pack.

Opened https://github.com/Jab125/fabric/pull/1 that should fix this.

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java (hunk 4 above)
Comment (Jab125) [reply fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ResourceManagerHelperImpl.java]: ends up going over to the hidden second line if the modid is too long

**Comment by Jab125** [comment]
> > This pull request as-is will break built-in resource packs that are not enabled by default on servers as they will be made loading by default. Reason being the mixin injection relies on reference equality of the source of the resource pack.
> 
> Opened [Jab125#1](https://github.com/Jab125/fabric/pull/1) that should fix this.

merged

**Comment by Jab125** [comment]
Should this pull request target 1.18.2?

**Comment by Player3324** [comment]
Yes, that's be great - sorry for the detail, we missed that the residual issues seem to have been fixed already

**Review by Player3324 — APPROVED** [review]

**Comment by Jab125** [comment]
Also, what do we do with the pack versions, as data-packs are version 9, while resource packs are version 8.
As we can create a pack with data and assets, shouldn't we be able to separate data and assets in the `pack_version`:
Example:
```{
  "pack": {
    "fabric:data_format": 9,
    "fabric:assets_format" 8,
    "description": "Fabric Resource Loader Test Builtin Pack."
  }
}```

**Review by PseudoDistant — APPROVED** [review]

**Comment by modmuss50** [comment]
Sorry for the delay on getting this merged, ive been busy with IRl things. I should be able to get this in very soon as things have started to settle down.

**Pull request merged**